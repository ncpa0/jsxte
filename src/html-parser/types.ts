export type RendererHTMLAttributes = [string, string | undefined][];

export type HTMLElementStruct = {
  tag: string;
  children: JSX.Element[];
  attributes: RendererHTMLAttributes;
};

export type Rewrap<T extends object> = T extends infer OBJ
  ? {
      [K in keyof OBJ]: OBJ[K] extends infer O ? O : never;
    }
  : never;
