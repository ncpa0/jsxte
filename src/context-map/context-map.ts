export class ContextMap {
  public static create(): ContextMap {
    return new ContextMap();
  }

  public static clone(original: ContextMap): ContextMap {
    return new ContextMap(new Map(original.map));
  }

  private constructor(private readonly map: Map<symbol, unknown> = new Map()) {}

  /**
   * Retrieve the context data for the specified context. If the
   * context has never been set by any of this component
   * ancestors an error will be thrown.
   */
  public get<T>(ref: ContextDefinition<T>): T {
    const value = this.map.get(ref.id);

    if (value === undefined) {
      throw new Error(
        "Context not defined! Make sure the context is set before accessing it."
      );
    }

    return value as T;
  }

  /**
   * Partially update the state of the context data. Works only
   * for objects and can only be used if some context data is
   * already set beforehand.
   *
   * Updates to the context made with this method are only
   * visible to this component and it's descendants.
   */
  public update<T extends object>(
    ref: ContextDefinition<T>,
    updateData: Partial<T>
  ): void {
    const data = this.get(ref);

    if (
      typeof data === "object" &&
      data !== null &&
      typeof updateData === "object" &&
      updateData !== null
    ) {
      if (Array.isArray(data)) {
        const arr = Array.from(data);

        for (const [key, value] of Object.entries(updateData)) {
          const index = Number(key);
          if (!isNaN(index)) arr[index] = value;
        }

        return void this.map.set(ref.id, arr);
      } else {
        return void this.map.set(ref.id, { ...data, ...updateData });
      }
    } else {
      throw new Error(
        "Context data is not an object!. Partial updates are only possible for objects."
      );
    }
  }

  /**
   * Sets the context data for the specified context.
   *
   * Changes to the context made with this method are only
   * visible to this component and it's descendants.
   */
  public set<T>(ref: ContextDefinition<T>, data: T): void {
    this.map.set(ref.id, data);
  }

  /** Check if the context data for the specified context is set. */
  public has<T>(ref: ContextDefinition<T>): boolean {
    return this.map.has(ref.id);
  }
}

export class ContextDefinition<T> {
  id = Symbol();
}

export const defineContext = <T = unknown>() => new ContextDefinition<T>();
