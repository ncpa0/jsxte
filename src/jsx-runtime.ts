export * from "./jsx/jsx.types";
export * from "./jsx/jsx-runtime";
