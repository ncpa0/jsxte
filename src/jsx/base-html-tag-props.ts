import type { Rewrap } from "../html-parser/types";

export type AttributeBool = true | false | "true" | "false";

export type HTMLProps<T extends object = never> = Rewrap<
  ExtendBaseProps<
    [T] extends [never]
      ? JSXTE.BaseHTMLTagProps
      : Partial<T> & JSXTE.BaseHTMLTagProps
  >
>;

type ExtendBaseProps<P> = {
  [K in keyof P]: JSXTE.AttributeAcceptedTypes extends {
    [E in K]: infer T;
  }
    ? T | P[K]
    : P[K];
};

declare global {
  namespace JSXTE {
    interface AttributeAcceptedTypes {}

    export type TagElement = {
      type: "tag";
      tag:
        | string
        | ((props: ElementProps) => Element)
        | ((props: ElementProps) => Promise<Element>);
      props: ElementProps;
    };

    export type TextNodeElement = {
      type: "textNode";
      text: string;
    };

    export type SyncElement = TagElement | TextNodeElement;

    type ElementChildren =
      | JSX.Element
      | string
      | number
      | Array<
          JSX.Element | string | number | Array<JSX.Element | string | number>
        >;

    type ElementProps = {
      children?: ElementChildren;
      [k: string]: any;
    };

    type PropsWithChildren<P extends object> = P & {
      children?: JSXTE.ElementChildren;
    };

    type Component<P extends object = {}> = (
      props: PropsWithChildren<P>
    ) => JSX.Element;

    type AsyncComponent<P extends object = {}> = (
      props: PropsWithChildren<P>
    ) => Promise<JSX.Element>;

    export interface BaseHTMLTagProps {
      children?: ElementChildren;

      accesskey?: string;
      class?: string;
      contenteditable?: AttributeBool;
      dir?: "ltr" | "rtl" | "auto";
      draggable?: AttributeBool | "auto";
      hidden?: AttributeBool;
      id?: string;
      inert?: AttributeBool;
      lang?: string;
      onabort?: string;
      onafterprint?: string;
      onbeforeprint?: string;
      onbeforeunload?: string;
      onblur?: string;
      oncanplay?: string;
      oncanplaythrough?: string;
      onchange?: string;
      onclick?: string;
      oncontextmenu?: string;
      oncopy?: string;
      oncuechange?: string;
      oncut?: string;
      ondblclick?: string;
      ondrag?: string;
      ondragend?: string;
      ondragenter?: string;
      ondragleave?: string;
      ondragover?: string;
      ondragstart?: string;
      ondrop?: string;
      ondurationchange?: string;
      onemptied?: string;
      onended?: string;
      onerror?: string;
      onfocus?: string;
      onhashchange?: string;
      oninput?: string;
      oninvalid?: string;
      onkeydown?: string;
      onkeypress?: string;
      onkeyup?: string;
      onload?: string;
      onloadeddata?: string;
      onloadedmetadata?: string;
      onloadstart?: string;
      onmousedown?: string;
      onmousemove?: string;
      onmouseout?: string;
      onmouseover?: string;
      onmouseup?: string;
      onmousewheel?: string;
      onoffline?: string;
      ononline?: string;
      onpagehide?: string;
      onpageshow?: string;
      onpaste?: string;
      onpause?: string;
      onplay?: string;
      onplaying?: string;
      onpopstate?: string;
      onprogress?: string;
      onratechange?: string;
      onreset?: string;
      onresize?: string;
      onscroll?: string;
      onsearch?: string;
      onseeked?: string;
      onseeking?: string;
      onselect?: string;
      onstalled?: string;
      onstorage?: string;
      onsubmit?: string;
      onsuspend?: string;
      ontimeupdate?: string;
      ontoggle?: string;
      onunload?: string;
      onvolumechange?: string;
      onwaiting?: string;
      onwheel?: string;
      role?: string;
      slot?: string;
      spellcheck?: AttributeBool;
      style?: string;
      tabindex?: string | number;
      title?: string;
      translate?: "yes" | "no";
    }
  }
}
