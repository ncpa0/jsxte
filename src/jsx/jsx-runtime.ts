import type { ComponentApi } from "../component-api/component-api";

type CreateElementProps = {
  [k: string]: any;
  children?: JSXTE.ElementChildren;
};

(() => {
  if(typeof Symbol.toHtmlTag === "symbol") return;

  Object.defineProperty(Symbol, "toHtmlTag", {
    value: Symbol("toHtmlTag"),
    enumerable: false,
    configurable: true,
  });
})();

declare global {
  interface SymbolConstructor {
    /**
     * Method under this property defines how an object should be converted to
     * HTML by the JSXTE renderer.
     *
     * @example
     *   class User {
     *     constructor(
     *       public firstname: string,
     *       public lastname: string,
     *       public email: string,
     *       public age: number,
     *     ) {}
     *
     *     [Symbol.toHtmlTag]() {
     *       return `${this.firstname} ${this.lastname}`;
     *     }
     *   }
     *
     *   const user = new User("John", "Doe", "joedoe@gmai.com", 26);
     *   renderToHtml(<div>{user}</div>); // => <div>John Doe</div>
     */
    readonly toHtmlTag: symbol;
  }
}

const mapChildren = (
  children: JSXTE.ElementChildren,
  accumulator: JSX.Element[],
) => {
  // eslint-disable-next-line @typescript-eslint/switch-exhaustiveness-check
  switch (typeof children) {
    case "string":
      accumulator.push({ type: "textNode", text: children });
      break;
    case "number":
      accumulator.push({ type: "textNode", text: children.toString() });
      break;
    case "object":
      if (Array.isArray(children)) {
        for (let i = 0; i < children.length; i++) {
          const child = children[i]!;
          mapChildren(child, accumulator);
        }
      } else if (children != null) {
        if (
          Symbol.toHtmlTag in children &&
          typeof children[Symbol.toHtmlTag] === "function"
        ) {
          const html = String((children[Symbol.toHtmlTag] as Function)());
          accumulator.push({ type: "textNode", text: html });
        } else {
          accumulator.push(children);
        }
      }
      break;
  }

  return accumulator;
};

export const createElement = (
  tag:
    | string
    | ((props: any, contextMap: ComponentApi) => JSX.Element)
    | ((props: any, contextMap: ComponentApi) => Promise<JSX.Element>),
  props?: CreateElementProps,
  ...children: Array<
    JSX.Element | string | number | Array<JSX.Element | string | number>
  >
): JSX.Element => {
  props ??= {};

  const finalChildren: JSX.Element[] = [];

  for (let i = 0; i < children.length; i++) {
    mapChildren(children[i]!, finalChildren);
  }

  if (props?.children) {
    mapChildren(props.children, finalChildren);
  }

  props.children = finalChildren;

  Object.freeze(finalChildren);
  Object.freeze(props);

  return {
    type: "tag",
    // @ts-expect-error
    tag,
    props: props as JSXTE.ElementProps,
  };
};

export const jsx = createElement;
export const jsxs = jsx;
export const _jsx = jsx;
export const _jsxs = jsx;

export const Fragment = "";
export const _Fragment = Fragment;
