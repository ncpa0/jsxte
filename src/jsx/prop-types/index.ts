export type { AnchorTagProps } from "./a-jsx-props";
export type { AreaTagProps } from "./area-jsx-props";
export type { AudioTagProps } from "./audio-jsx-props";
export type { BaseTagProps } from "./base-jsx-props";
export type { BdoTagProps } from "./bdo-jsx-props";
export type { BlockquoteTagProps } from "./blockquote-jsx-props";
export type { ButtonTagProps } from "./button-jsx-props";
export type { CanvasTagProps } from "./canvas-jsx-props";
export type { ColTagProps } from "./col-jsx-props";
export type { ColgroupTagProps } from "./colgroup-jsx-props";
export type { DataTagProps } from "./data-jsx-props";
export type { DelTagProps } from "./del-jsx-props";
export type { DetailsTagProps } from "./details-jsx-props";
export type { DialogTagProps } from "./dialog-jsx-props";
export type { EmbedTagProps } from "./embed-jsx-props";
export type { FieldsetTagProps } from "./fieldset-jsx-prop";
export type { FormTagProps } from "./form-jsx-props";
export type { HtmlTagProps } from "./html-jsx-props";
export type { IframeTagProps } from "./iframe-jsx-props";
export type { ImgTagProps } from "./img-jsx-props";
export type { InputTagProps } from "./input-jsx-props";
export type { InsTagProps } from "./ins-jsx-props";
export type { LabelTagProps } from "./label-jsx-props";
export type { LiTagProps } from "./li-jsx.props";
export type { LinkTagProps } from "./link-jsx-props";
export type { MapTagProps } from "./map-jsx-props";
export type { MetaTagProps } from "./meta-jsx-props";
export type { MeterTagProps } from "./meter-jsx-props";
export type { ObjectTagProps } from "./object-jsx-props";
export type { OlTagProps } from "./ol-jsx-props";
export type { OptgroupTagProps } from "./optgroup-jsx-props";
export type { OptionTagProps } from "./option-jsx-props";
export type { OutputTagProps } from "./output-jsx-props";
export type { ParamTagProps } from "./param-jsx-props";
export type { ProgressTagProps } from "./progress-jsx-props";
export type { QTagProps } from "./q-jsx-props";
export type { ScriptTagProps } from "./script-jsx-props";
export type { SelectTagProps } from "./select-jsx-props";
export type { SourceTagProps } from "./source-jsx-props";
export type { StyleTagProps } from "./style-jsx-props";
export type { TdTagProps } from "./td-jsx-props";
export type { TextareaTagProps } from "./textarea-jsx-props";
export type { ThTagProps } from "./th-jsx-props";
export type { TimeTagProps } from "./time-jsx-props";
export type { TrackTagProps } from "./track-jsx-props";
export type { VideoTagProps } from "./video-jsx-props";
