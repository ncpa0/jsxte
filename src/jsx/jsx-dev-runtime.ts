import type { ComponentApi } from "../component-api/component-api";
import type { CreateElementProps } from "./jsx-runtime";
import { Fragment, jsx, jsxs } from "./jsx-runtime";

const jsxDEV = (
  tag:
    | string
    | ((props: any, contextMap: ComponentApi) => JSX.Element)
    | ((props: any, contextMap: ComponentApi) => Promise<JSX.Element>),
  props?: CreateElementProps,
) => {
  const { children, ...restProps } = props ?? {};
  if (Array.isArray(children)) {
    return jsx(tag, restProps, ...children);
  }
  return jsx(tag, restProps, children);
};

export { Fragment, jsx, jsxs, jsxDEV };
