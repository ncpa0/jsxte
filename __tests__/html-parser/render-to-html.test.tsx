import "../../src/jsx/jsx.types";
import {
  renderToHtml,
  renderToHtmlAsync,
} from "../../src/html-parser/render-to-html";
// @ts-ignore
import { jsx, Fragment } from "../../src/jsx/jsx-runtime";
import { defineContext } from "../../src/context-map/context-map";

const sleep = (t: number) =>
  new Promise<void>((resolve) => setTimeout(() => resolve(), t));

describe("renderToHTML", () => {
  it("should correctly generate html from simple jsx", () => {
    const Component = (props: { title: string }) => {
      return (
        <div id="container" class={"bordered active"}>
          <h1>Hello World</h1>
          <h2>{props.title}</h2>
          <button onclick={'console.log("Hello World!")'}>Click me!</button>
          <input autofocus={true} disabled={false} draggable />
        </div>
      );
    };

    const html = renderToHtml(<Component title="Prop Title" />);

    expect(html).toMatchSnapshot();
  });

  it("should correctly generate html from component base jsx structure", () => {
    const Header = (props: { title: string }) => {
      return <h2>{props.title}</h2>;
    };

    const Button = (props: { label: string; styles?: string }) => {
      return <button style={props.styles}>{props.label}</button>;
    };

    const Template = ({ children }: JSXTE.ElementProps) => {
      return (
        <html>
          <head>
            <meta charset="utf-8" />
            <meta http-equiv="x-ua-compatible" content="IE=edge" />
            <title>Page Title</title>
            <meta
              name="viewport"
              content="width=device-width, initial-scale=1"
            />
            <link
              rel="stylesheet"
              type="text/css"
              media="screen"
              href="main.css"
            />
          </head>
          <body>{children}</body>
        </html>
      );
    };

    const Content = () => {
      return (
        <>
          <Header title="Hello World!" />
          <input value="write here" />
          <Button label="Submit" />
        </>
      );
    };

    const App = () => {
      return (
        <Template>
          <div class="main-container">
            <Content />
          </div>
        </Template>
      );
    };

    const structure = <App />;

    const html = renderToHtml(structure);

    expect(html).toMatchSnapshot();
  });

  it("should correctly render jsx with arrays in between elements", () => {
    const App = () => {
      return (
        <div>
          <h1>Header</h1>
          {Array.from({ length: 2 }, (_, i) => i).map((i) => (
            <p>{i}</p>
          ))}
          <footer />
        </div>
      );
    };

    const html = renderToHtml(<App />);

    expect(html).toMatchSnapshot();
  });

  it("should correctly parse async components", async () => {
    const Header = async (props: { title: string }) => {
      await sleep(50);
      return <h2>{props.title}</h2>;
    };

    const Button = async (props: { label: string; styles?: string }) => {
      await sleep(50);
      return <button style={props.styles}>{props.label}</button>;
    };

    const Template = ({ children }: JSXTE.ElementProps) => {
      return (
        <html>
          <head>
            <meta charset="utf-8" />
            <meta http-equiv="x-ua-compatible" content="IE=edge" />
            <title>Page Title</title>
            <meta
              name="viewport"
              content="width=device-width, initial-scale=1"
            />
            <link
              rel="stylesheet"
              type="text/css"
              media="screen"
              href="main.css"
            />
          </head>
          <body>{children}</body>
        </html>
      );
    };

    const Content = async () => {
      const headerTitle = await sleep(100).then(() => "Async title");
      return (
        <>
          <Header title={headerTitle} />
          <input value="write here" />
          <Button label="Submit" />
        </>
      );
    };

    const App = () => {
      return (
        <Template>
          <div class="main-container">
            <Content />
          </div>
        </Template>
      );
    };

    expect(() => renderToHtml(<App />)).toThrowError();

    const html = await renderToHtmlAsync(<App />);
    expect(html).toMatchSnapshot();
  });

  it("should correctly handle nulls, undefined and boolean values", () => {
    const StandardDiv = () => {
      return (
        <div>
          <div></div>
          <div></div>
          <div>{[<span>before</span>, <span>after</span>]}</div>
        </div>
      );
    };

    const DivWithNulls = () => {
      return (
        <div>
          <div>{null}</div>
          <div>{[null]}</div>
          <div>{[<span>before</span>, null, null, <span>after</span>]}</div>
        </div>
      );
    };

    const DivWithUndefined = () => {
      return (
        <div>
          <div>{undefined}</div>
          <div>{[undefined]}</div>
          <div>
            {[<span>before</span>, undefined, undefined, <span>after</span>]}
          </div>
        </div>
      );
    };

    const DivWithFalse = () => {
      return (
        <div>
          <div>{false}</div>
          <div>{[false]}</div>
          <div>{[<span>before</span>, false, false, <span>after</span>]}</div>
        </div>
      );
    };

    const DivWithTrue = () => {
      return (
        <div>
          <div>{true}</div>
          <div>{[true]}</div>
          <div>{[<span>before</span>, true, true, <span>after</span>]}</div>
        </div>
      );
    };

    const renderedStandardDiv = renderToHtml(<StandardDiv />);

    const renderedDivWithNulls = renderToHtml(<DivWithNulls />);
    const renderedDivWithUndefined = renderToHtml(<DivWithUndefined />);
    const renderedDivWithFalse = renderToHtml(<DivWithFalse />);
    const renderedDivWithTrue = renderToHtml(<DivWithTrue />);

    expect(renderedStandardDiv).toEqual(renderedDivWithNulls);
    expect(renderedStandardDiv).toEqual(renderedDivWithUndefined);
    expect(renderedStandardDiv).toEqual(renderedDivWithFalse);
    expect(renderedStandardDiv).toEqual(renderedDivWithTrue);
  });
  describe("should properly handle context data", () => {
    it("should correctly render jsx with context data", () => {
      const context = defineContext<{ title: string }>();

      const Header: JSXTE.Component = (_, contextMap) => {
        expect(contextMap.has(context)).toBe(true);
        const { title } = contextMap.get(context);
        expect(title).toBe("This title is set via the context");
        return (
          <div>
            <h1>{title}</h1>
          </div>
        );
      };

      const Content: JSXTE.Component = () => {
        return (
          <div class="main-container">
            <Header />
          </div>
        );
      };

      const App: JSXTE.Component = (_, contextMap) => {
        contextMap.set(context, { title: "This title is set via the context" });

        return (
          <html>
            <head>
              <meta charset="utf-8" />
              <meta http-equiv="x-ua-compatible" content="IE=edge" />
              <title>Page Title</title>
              <meta
                name="viewport"
                content="width=device-width, initial-scale=1"
              />
              <link
                rel="stylesheet"
                type="text/css"
                media="screen"
                href="main.css"
              />
            </head>
            <body>
              <Content />
            </body>
          </html>
        );
      };

      const html = renderToHtml(<App />);

      expect(html).toMatchSnapshot();
    });

    it("should correctly render jsx with context data and arrays in between elements", () => {
      const context = defineContext<{
        title: string;
        inputPlaceholder: string;
        buttonLabel: string;
      }>();

      const Header: JSXTE.Component = (_, contextMap) => {
        expect(contextMap.has(context)).toBe(true);
        const { title } = contextMap.get(context);

        contextMap.set(context, {
          title,
          inputPlaceholder:
            "This should not affect the rendered content, since this component has no children that consume this context.",
          buttonLabel:
            "This should not affect the rendered content, since this component has no children that consume this context.",
        });

        return (
          <div>
            <h1>{title}</h1>
          </div>
        );
      };

      const Input: JSXTE.Component = (_, contextMap) => {
        expect(contextMap.has(context)).toBe(true);
        const { inputPlaceholder } = contextMap.get(context);
        expect(inputPlaceholder).toBe("write here");
        return <input placeholder={inputPlaceholder} />;
      };

      const Button: JSXTE.Component = (_, contextMap) => {
        expect(contextMap.has(context)).toBe(true);
        const { buttonLabel } = contextMap.get(context);
        expect(buttonLabel).toBe("Submit");
        return <button>{buttonLabel}</button>;
      };

      const Content: JSXTE.Component = () => {
        return <>{[<Header />, <Input />, <Button />]}</>;
      };

      const App: JSXTE.Component = (_, contextMap) => {
        contextMap.set(context, {
          title: "This title is set via the context",
          buttonLabel: "Submit",
          inputPlaceholder: "write here",
        });

        return (
          <html>
            <head>
              <meta charset="utf-8" />
              <meta http-equiv="x-ua-compatible" content="IE=edge" />
              <title>Page Title</title>
              <meta
                name="viewport"
                content="width=device-width, initial-scale=1"
              />
              <link
                rel="stylesheet"
                type="text/css"
                media="screen"
                href="main.css"
              />
            </head>
            <body>
              <Content />
            </body>
          </html>
        );
      };

      const html = renderToHtml(<App />);

      expect(html).toMatchSnapshot();
    });

    it("should correctly render jsx with context data and async components", async () => {
      const context = defineContext<{
        title: string;
        inputPlaceholder: string;
        buttonLabel: string;
      }>();

      const Header: JSXTE.Component = async (_, contextMap) => {
        expect(contextMap.has(context)).toBe(true);
        const { title } = contextMap.get(context);

        contextMap.set(context, {
          title,
          inputPlaceholder:
            "This should not affect the rendered content, since this component has no children that consume this context.",
          buttonLabel:
            "This should not affect the rendered content, since this component has no children that consume this context.",
        });

        await sleep(10);
        return (
          <div>
            <h1>{title}</h1>
          </div>
        );
      };

      const Input: JSXTE.Component = async (_, contextMap) => {
        expect(contextMap.has(context)).toBe(true);
        const { inputPlaceholder } = contextMap.get(context);
        expect(inputPlaceholder).toBe("write here");
        await sleep(50);
        return <input placeholder={inputPlaceholder} />;
      };

      const Button: JSXTE.Component = async (_, contextMap) => {
        expect(contextMap.has(context)).toBe(true);
        const { buttonLabel } = contextMap.get(context);
        expect(buttonLabel).toBe("Submit");
        await sleep(50);
        return <button>{buttonLabel}</button>;
      };

      const Content: JSXTE.Component = () => {
        return <>{[<Header />, <Input />, <Button />]}</>;
      };

      const App: JSXTE.Component = async (_, contextMap) => {
        contextMap.set(context, {
          title: "This title is set via the context",
          buttonLabel: "Submit",
          inputPlaceholder: "write here",
        });

        return (
          <html>
            <head>
              <meta charset="utf-8" />
              <meta http-equiv="x-ua-compatible" content="IE=edge" />
              <title>Page Title</title>
              <meta
                name="viewport"
                content="width=device-width, initial-scale=1"
              />
              <link
                rel="stylesheet"
                type="text/css"
                media="screen"
                href="main.css"
              />
            </head>
            <body>
              <Content />
            </body>
          </html>
        );
      };

      const html = await renderToHtmlAsync(<App />);

      expect(html).toMatchSnapshot();
    });

    it("should correctly override existing context data", async () => {
      const context = defineContext<{
        title: string;
        inputPlaceholder: string;
      }>();

      const Header: JSXTE.Component = async (_, contextMap) => {
        expect(contextMap.has(context)).toBe(true);
        const { title } = contextMap.get(context);
        expect(title).toBe("This title was overridden");
        return <h1>{title}</h1>;
      };

      const Input: JSXTE.Component = async (_, contextMap) => {
        expect(contextMap.has(context)).toBe(true);
        const { inputPlaceholder } = contextMap.get(context);
        expect(inputPlaceholder).toBe("write here");
        await sleep(50);
        return <input placeholder={inputPlaceholder} />;
      };

      const Content: JSXTE.Component = (_, contextMap) => {
        expect(contextMap.has(context)).toBe(true);
        const { title } = contextMap.get(context);
        expect(title).toBe("This title was set in the app component");

        contextMap.update(context, {
          title: "This title was overridden",
        });

        return (
          <>
            <Header />
            <Input />
          </>
        );
      };

      const App: JSXTE.Component = async (_, contextMap) => {
        contextMap.set(context, {
          title: "This title was set in the app component",
          inputPlaceholder: "write here",
        });

        return (
          <html>
            <head>
              <meta charset="utf-8" />
              <meta http-equiv="x-ua-compatible" content="IE=edge" />
              <title>Page Title</title>
              <meta
                name="viewport"
                content="width=device-width, initial-scale=1"
              />
              <link
                rel="stylesheet"
                type="text/css"
                media="screen"
                href="main.css"
              />
            </head>
            <body>
              <Content />
            </body>
          </html>
        );
      };

      const html = await renderToHtmlAsync(<App />);

      expect(html).toMatchSnapshot();
    });
  });
});
