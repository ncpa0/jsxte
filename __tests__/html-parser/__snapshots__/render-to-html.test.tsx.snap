// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renderToHTML should correctly generate html from component base jsx structure 1`] = `
"<html>
  <head>
    <meta charset=\\"utf-8\\">
    </meta>
    <meta http-equiv=\\"x-ua-compatible\\" content=\\"IE=edge\\">
    </meta>
    <title>
Page Title
    </title>
    <meta name=\\"viewport\\" content=\\"width=device-width, initial-scale=1\\">
    </meta>
    <link rel=\\"stylesheet\\" type=\\"text/css\\" media=\\"screen\\" href=\\"main.css\\">
    </link>
  </head>
  <body>
    <div class=\\"main-container\\">
        <h2>
Hello World!
        </h2>
        <input value=\\"write here\\">
        </input>
        <button>
Submit
        </button>
    </div>
  </body>
</html>"
`;

exports[`renderToHTML should correctly generate html from simple jsx 1`] = `
"<div id=\\"container\\" class=\\"bordered active\\">
  <h1>
Hello World
  </h1>
  <h2>
Prop Title
  </h2>
  <button onclick=\\"console.log(&quot;Hello World!&quot;)\\">
Click me!
  </button>
  <input autofocus draggable>
  </input>
</div>"
`;

exports[`renderToHTML should correctly parse async components 1`] = `
"<html>
  <head>
    <meta charset=\\"utf-8\\">
    </meta>
    <meta http-equiv=\\"x-ua-compatible\\" content=\\"IE=edge\\">
    </meta>
    <title>
Page Title
    </title>
    <meta name=\\"viewport\\" content=\\"width=device-width, initial-scale=1\\">
    </meta>
    <link rel=\\"stylesheet\\" type=\\"text/css\\" media=\\"screen\\" href=\\"main.css\\">
    </link>
  </head>
  <body>
    <div class=\\"main-container\\">
        <h2>
Async title
        </h2>
        <input value=\\"write here\\">
        </input>
        <button>
Submit
        </button>
    </div>
  </body>
</html>"
`;

exports[`renderToHTML should correctly render jsx with arrays in between elements 1`] = `
"<div>
  <h1>
Header
  </h1>
  <p>
0
  </p>
  <p>
1
  </p>
  <footer>
  </footer>
</div>"
`;

exports[`renderToHTML should properly handle context data close-by providers should not interfere with each other 1`] = `
"<html>
  <body>
        <p>
bar
        </p>
        <p>
baz
        </p>
        <p>
qux
        </p>
        <p>
coorg
        </p>
        <p>
fred
        </p>
        <p>
thud
        </p>
      <p>
foo
      </p>
  </body>
</html>"
`;

exports[`renderToHTML should properly handle context data correctly handles encapsulated providers 1`] = `
"<html>
  <body>
    <div id=\\"root\\">
        <div class=\\"template for-magic\\">
          <div>
              <input type=\\"text\\" name=\\"repeating_magic_dwarf\\">
              </input>
          </div>
        </div>
    </div>
  </body>
</html>"
`;

exports[`renderToHTML should properly handle context data should correctly drill the prop data through multiple components 1`] = `
"<html>
  <body>
      <div>
        <div>
          <div>
            <div>
              <div>
                <div>
                  <div>
                    <div>
                      <p>
foo
                      </p>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
  </body>
</html>"
`;

exports[`renderToHTML should properly handle context data should correctly handle provider pattern 1`] = `
"<html>
  <head>
    <meta charset=\\"utf-8\\">
    </meta>
    <meta http-equiv=\\"x-ua-compatible\\" content=\\"IE=edge\\">
    </meta>
    <title>
Page Title
    </title>
    <meta name=\\"viewport\\" content=\\"width=device-width, initial-scale=1\\">
    </meta>
    <link rel=\\"stylesheet\\" type=\\"text/css\\" media=\\"screen\\" href=\\"main.css\\">
    </link>
  </head>
    <body>
      <div class=\\"main-container\\">
        <div>
          <h1>
Provided title
          </h1>
        </div>
      </div>
    </body>
</html>"
`;

exports[`renderToHTML should properly handle context data should correctly override existing context data 1`] = `
"<html>
  <head>
    <meta charset=\\"utf-8\\">
    </meta>
    <meta http-equiv=\\"x-ua-compatible\\" content=\\"IE=edge\\">
    </meta>
    <title>
Page Title
    </title>
    <meta name=\\"viewport\\" content=\\"width=device-width, initial-scale=1\\">
    </meta>
    <link rel=\\"stylesheet\\" type=\\"text/css\\" media=\\"screen\\" href=\\"main.css\\">
    </link>
  </head>
  <body>
      <h1>
This title was overridden
      </h1>
      <input placeholder=\\"write here\\">
      </input>
  </body>
</html>"
`;

exports[`renderToHTML should properly handle context data should correctly render jsx with context data 1`] = `
"<html>
  <head>
    <meta charset=\\"utf-8\\">
    </meta>
    <meta http-equiv=\\"x-ua-compatible\\" content=\\"IE=edge\\">
    </meta>
    <title>
Page Title
    </title>
    <meta name=\\"viewport\\" content=\\"width=device-width, initial-scale=1\\">
    </meta>
    <link rel=\\"stylesheet\\" type=\\"text/css\\" media=\\"screen\\" href=\\"main.css\\">
    </link>
  </head>
  <body>
    <div class=\\"main-container\\">
      <div>
        <h1>
This title is set via the context
        </h1>
      </div>
    </div>
  </body>
</html>"
`;

exports[`renderToHTML should properly handle context data should correctly render jsx with context data and arrays in between elements 1`] = `
"<html>
  <head>
    <meta charset=\\"utf-8\\">
    </meta>
    <meta http-equiv=\\"x-ua-compatible\\" content=\\"IE=edge\\">
    </meta>
    <title>
Page Title
    </title>
    <meta name=\\"viewport\\" content=\\"width=device-width, initial-scale=1\\">
    </meta>
    <link rel=\\"stylesheet\\" type=\\"text/css\\" media=\\"screen\\" href=\\"main.css\\">
    </link>
  </head>
  <body>
      <div>
        <h1>
This title is set via the context
        </h1>
      </div>
      <input placeholder=\\"write here\\">
      </input>
      <button>
Submit
      </button>
  </body>
</html>"
`;

exports[`renderToHTML should properly handle context data should correctly render jsx with context data and async components 1`] = `
"<html>
  <head>
    <meta charset=\\"utf-8\\">
    </meta>
    <meta http-equiv=\\"x-ua-compatible\\" content=\\"IE=edge\\">
    </meta>
    <title>
Page Title
    </title>
    <meta name=\\"viewport\\" content=\\"width=device-width, initial-scale=1\\">
    </meta>
    <link rel=\\"stylesheet\\" type=\\"text/css\\" media=\\"screen\\" href=\\"main.css\\">
    </link>
  </head>
  <body>
      <div>
        <h1>
This title is set via the context
        </h1>
      </div>
      <input placeholder=\\"write here\\">
      </input>
      <button>
Submit
      </button>
  </body>
</html>"
`;
